packages/viewport/src/with-viewport-match.js

```diff
@@ -1,13 +1,83 @@
 /**
  * External dependencies
  */
-import { mapValues } from 'lodash';
+import { mapValues, reduce } from 'lodash';
 
 /**
  * WordPress dependencies
  */
+import { createContext, useContext, useMemo } from '@wordpress/element';
 import { createHigherOrderComponent } from '@wordpress/compose';
-import { withSelect } from '@wordpress/data';
+import { useSelect } from '@wordpress/data';
+
+/**
+ * Hash of breakpoint names with pixel width at which it becomes effective.
+ *
+ * @see _breakpoints.scss
+ *
+ * @type {Object}
+ */
+export const BREAKPOINTS = {
+	huge: 1440,
+	wide: 1280,
+	large: 960,
+	medium: 782,
+	small: 600,
+	mobile: 480,
+};
+
+/**
+ * Hash of query operators with corresponding condition for media query.
+ *
+ * @type {Object}
+ */
+export const OPERATORS = {
+	'<': 'max-width',
+	'>=': 'min-width',
+};
+
+const OPERATOR_EVALUATORS = {
+	'<': ( breakpointValue, width ) => ( width < breakpointValue ),
+	'>=': ( breakpointValue, width ) => ( width >= breakpointValue ),
+};
+
+const ViewportMatchWidthContext = createContext( null );
+
+/**
+ * Component that makes all withViewportMatch usages descendents of its children,
+ * evaluate as if the viewport had the width specific as a prop.
+ */
+export const ViewportMatchWidthProvider = ( { width, children } ) => {
+	const queriesResult = useMemo(
+		() => {
+			return reduce(
+				BREAKPOINTS,
+				( resultBreakpointsReduce, breakpointValue, breakpointName ) => {
+					return reduce(
+						OPERATORS,
+						( resultOperatorsReduce, operatorMediaQuery, operator ) => {
+							const operatorKey = `${ operator } ${ breakpointName }`;
+							const evaluationResult = OPERATOR_EVALUATORS[ operator ](
+								breakpointValue,
```

Comment:
This nesting gets pretty excessive and becomes hard to follow.

I might suggest:

- Not every argument needs to be on its own line
- Consolidate the operations.

That being said, since we're combining an aggregate of two different objects, I guess it's inevitable that we'll have some nesting like this.